lib/bigcommerce/prometheus/integrations/railtie.rb

```diff
@@ -22,7 +22,7 @@ module Integrations
       # Railtie for automatic configuration of Rails environments
       #
       class Railtie < ::Rails::Railtie
-        initializer 'zzz.bc_prometheus_ruby.configure_rails_initialization' do |app|
+        config.after_initialize do |app|
```

Comment:
This will not work as you expect. OOE of initializers changes in Rails 5/6 (even in 5.0/5.1/5.2!), and this will cause issues with other systems as Rails freezes its middleware, causing deadlocks and crashes.

Reply:
Good to know. I haven't seen syntax like that before. Can you provide a link/description of how it works? I'm assuming we're relying on `zzz` to try and make it load later on at least.

In that case, there's a good chance there's something misconfigured in our service then yeah? As best I can tell, we're still following the setup instructions on the readme. I'll double check soon.

Reply:
Apologies for the delay here. We were impacted by the crazy Texas weather.

Disclaimer: I'm learning about the Rails boot order/etc as I go. I very well might have interpreted the docs incorrectly 👨‍🎓 

---

Here's a simple set of findings from local debugging:

`initializer 'zzz.bc_prometheus_ruby.configure_rails_initialization' do |app|`
- runs BEFORE our app's `config/initializers/01_instrumentation.rb`
- custom app metrics do not work with OOTB v0.4.0 `bc-prometheus-ruby`

`config.after_initialize do |app|`
- runs AFTER our app's `config/initializers/01_instrumentation.rb`
- custom app metrics work with modified v0.4.0 `bc-prometheus-ruby`
  - modified == this PR

I did some looking into how Rails handles loading "all the things." My two big takeaways:

Rails uses `tsort_each` under the hood to determine initializer ordering:
- https://guides.rubyonrails.org/v5.2/initialization.html#railties-lib-rails-application-rb
- https://github.com/rails/rails/blob/v5.2.4.5/railties/lib/rails/initializable.rb#L46-L64
- https://apidock.com/ruby/TSort/tsort_each


> Initializers defined using the initializer method will be run in the order they are defined in, with the exception of ones that use the :before or :after methods.
- https://guides.rubyonrails.org/configuring.html#rails-railtie-initializer

The way I read this, `zzz.bc_prometheus_ruby.configure_rails_initialization` should help us out on the `initialize!` / `run_initializers` call that uses `tsort_each`. The `:after` method should run after all that has happened. I _think_ this is backed up by `Bigcommerce::Prometheus::Instrumentors::Web.new(app: app).start` not reporting on our custom app metrics until after our app initializes everything else [0].

Could something like this be a viable solution?

```
initializer 'zzz.bc_prometheus_ruby.configure_rails_initialization' do
end

config.after_initialize do |app|
  Bigcommerce::Prometheus::Instrumentors::Web.new(app: app).start
end
```

With this, we'll help `tsort_each` initialize `bc-prometheus-ruby` as late as possible [1] and engage app's custom metrics after everything has booted up.

---

[0] - I haven't yet found the _exact_ line `start` can be called and work yet
[1] - I'm not sure this actually does anything different given the block is effectively a noop 😞

Reply:
Hm, could we get the best of both worlds by doing something like this instead @splittingred ?

```
initializer 'zzz.bc_prometheus_ruby.configure_rails_initialization' do |app|
  config.after_initialize do |app|
    Bigcommerce::Prometheus::Instrumentors::Web.new(app: app).start
  end
end
```

Reply:
I don't think that'd work - I think you'd end up causing it not to fire in some Rails versions.

@jcolbyfisher I think the best thing here for you to do is try an approach that is normal - say, what's in this PR - and then try it on the following systems:

* A&A
* BMP
* Launchbay
* SSL Service

Each of those have varying Rails versions, and you should be able to start with `PROMETHEUS_ENABLED=1` in .env and then hit the metrics endpoint to confirm they worked. Launchbay and BMP have custom type collectors as well, which you'll want to confirm those load.

Reply:
I can confirm a simpler service that also uses Rails 5.4.2.5 works with OOTB with 0.4.0. I suspect other services will likely work as well. I'll see about checking them out.

There's gotta be something weird going on with our favorite large service, and we haven't quite figured it out yet. Hopefully I'll have more to report at a later time.